src/crimson/osd/pg.cc

```diff
@@ -764,6 +764,10 @@ PG::submit_transaction(
   peering_state.append_log_with_trim_to_updated(std::move(log_entries), osd_op_p.at_version,
 						txn, true, false);
 
+  ceph_assert(!log_entries.empty());
+  ceph_assert(log_entries.rbegin()->version >= projected_last_update);
+  projected_last_update = log_entries.rbegin()->version;
```

Comment:
Yeah, backfill has the `PGFacade::get_projected_last_update()` junction point.

What makes be a bit worry is it's once-and-for-all fix for all places where `next_version()` has been missed. On the other hand, `submit_error_log()` already does very similar thing.

Reply:
> What makes be a bit worry is it's once-and-for-all fix for all places where `next_version()` has been missed. On the other hand, `submit_error_log()` already does very similar thing.

Should we remove those lines from `submit_error_log()`?

Reply:
Yeah, if next_version() exists in its current form, it must be the *only* place where projected_last_update changes.  If we want to go this direction, let's modify submit_error_log to work the same way and remove next_version().

Reply:
> Yeah, if next_version() exists in its current form, it must be the _only_ place where projected_last_update changes. If we want to go this direction, let's modify submit_error_log to work the same way and remove next_version().

Actually, I'm a bit confused about why `PG::next_version()` doesn't update `PG::projected_last_update`'s epoch. It seems that this is modeling what classic OSD is doing, but why?

Reply:
PrimaryLogPG::issue_repop does the update of projected_last_update, basically like you are doing in this PR.  PrimaryLogPG::get_next_version() is just a helper to get the numerically next version.

Reply:
The approach you are doing here and that classic does is, to me, the better approach.  It couples the update of projected_last_update with the actual submission of the op.

Reply:
Got it:-)

Reply:
@athanatos BTW, for the same repop, all pg logs should have the same version, am I right?

I'm asking this because it seems that there exist circumstances in which the same client/snap_trim requests calls more than once `PG::next_version()`, and the current implementation `PG::next_version()` does update `PG::projected_last_update`'s version, so the result of later invocation of `PG::next_version()` would be larger than earlier ones, which means there may be pg logs of the same repop having different versions.

Reply:
If a repop is submitting two different log entries, they should have successive versions.  Upon submission, projected_last_update should be the same as the last one.  Calling next_version() twice for two log entries would have the correct behavior in that case.

Reply:
Got it.

Reply:
Done:-)